Duplicati/Library/Main/Operation/DeleteHandler.cs

```diff
@@ -221,34 +228,41 @@ private DateTime[] GetFilesetsToDelete(Database.LocalDeleteDatabase db, DateTime
             var fullVersionsKeptCount = 0;
             if (fullVersionsToKeep > 0 && fullVersionsToKeep < backupsRemaining.Count)
             {
-                // keep the number of full backups specified in fullVersionsToKeep.
-                // once the last full backup t okeep is found, also keep the partials immediately after it the full backup.
-                // add the remainder of full and partial backups to toDelete
-                bool foundLastFullBackupToKeep = false;
+                ISet<DateTime> intermediatePartials = new HashSet<DateTime>();

```

Comment:
I think it is good to split this function in a part for keep time and keep version. I think it is excellent you wrote tests for them. If the logic is contained in separate functions, one can write tests just against such functions instead of going through the entire duplicati stack with controller etc. to reach them.

Reply:
I think that's a great idea.  However, it sounds like a larger refactoring that might require rewriting the tests to create entries in the database without actually running the backups.  If it's ok with you, I'd like to leave this for a future pull request.

Going through the controller does have a slight appeal to me, as it exercises the handling of the cancellation tokens...